BIP 79: Bustapay :: a practical coinjoin protocol

RHavar opened the pull request:
Requesting a BIP number assignment

luke-jr commented on bip-bustapay.mediawiki:
```diff
@@ -0,0 +1,120 @@
+<pre>
+  BIP: 69
```
Please don't make up BIP numbers.

luke-jr commented on bip-bustapay.mediawiki:
Should the sender verify that its change has not increased? That might be a confusing situation for the user.

RHavar replied:
Probably not a big problem, I've updated the "output adjustment" section to actually allow this situation -- as I just don't see it as a problem. And it might be kind of useful if a receiver wants to incentivize or reimburse users using bustapay (like refunding a "consolidation fee" that some services like bitpay put on the invoice)

luke-jr replied:
Someone using BitPay won't be able to implement this BIP - only BitPay themselves could. And in that case, they can just not add the fee...

RHavar replied:
When bitpay creates an invoice they won't know if you will pay via a normal transaction or bustpay. They have to always assume that a normal transaction will be used to make the payment, so the invoice amount will reflect that. But if the invoice is finally paid via bustpay, they could potentially offer a discount (if they wanted)

luke-jr commented on bip-bustapay.mediawiki:
What happens if both the sender and receiver have the same private key in their wallets?

RHavar replied:
Not really a problem. In the "reference implementation" it does support using the same wallet for both sending and receiving. As long as you verify the inputs were in the template transaction, as opposed to checking if you "own" them (like it says here).

luke-jr replied:
The receiver-contributed input *won't* have been in the template transaction.

luke-jr commented on bip-bustapay.mediawiki:
First sentence is nonsensical?

luke-jr commented on bip-bustapay.mediawiki:
Since the receiver is the most interested party to the transaction confirming, it would seem ideal for a norm where the sender does not contribute any fees (or perhaps only fees covering its inputs) and the receiver contributes whatever fees they desire.

luke-jr replied:
I wonder if it might be appropriate to use SIGHASH_SINGLE or such so that the receiver can add its input/output *and* propagate the new transaction without going back through the sender...

RHavar replied:
That would be more robust against spies/DoS, but would make it obvious what is a bustapayment transactions and thus negate the primary privacy benefits bustapay offers

luke-jr commented on bip-bustapay.mediawiki:
Why not drop the address, and have the server add the output? This enables the recipient to use any output it wants (not limited to merely address-representable outputs the sender already can understand; eg, it could use a native segwit output even with senders that don't support Bech32), as well as choose a key per-request (so the URI can be placed on a static webpage, yet still avoid address reuse).

RHavar replied:
The big reason is that if the receiver an arbitrarily change his output, the sender has no way to validate it. Any MITM  (imagine cloudflare) could change the output to their own, and then the sender would send them money. While with the current design, a MITM can only function as a DoS  (in which case the sender will just fall back to a normal send).

There was a suggestion that in the "payment request" it encodes what the "new output" would be changed to. However, I didn't think I could justify the complexity at this point in time.

luke-jr commented on bip-bustapay.mediawiki:
Why only segwit inputs?

RHavar replied:
Simplicity/txid malleability.

The receiver needs to watch the network for the final payment, and needs to differentiate a normal payment and a bustapayment to the same address. So if there's no transaction id malleability this is simpler as the final transaction will have the same txid as the partial transaction txid.

It definitely would be possible to support "legacy" transactions, just not worth the effort.

RHavar commented on bip-bustapay.mediawiki:
It is fully signed. This is important so the receiver can push the transaction onto the network (make it cost spending a utxo to learn one of the receivers utxos)

luke-jr replied:
As soon as a signed transaction is sent (even privately), we must assume the transaction has successfully completed, and cannot give the user the impression it failed.

RHavar commented on bip-bustapay.mediawiki:
It is not WebDav-compatible.  I find it's a lot nicer to use more specific error codes than 400, as something like 400 is frequently used by reverse-proxies and http libraries in general. 

If you want, I'll change it over to 400, but I think 422 is better.

luke-jr replied:
I suppose the BIP can itself define the 422 outside of the WebDAV context.

RHavar commented on bip-bustapay.mediawiki:
I meant to write "generally should not", i'll rework this whole paragraph

RHavar commented:
@luke-jr  I think I addressed most of your comments, except adding a backwards compatibility section -- as I'm not quite sure what I should put there. I don't think there's really any backwards compatibility issues

luke-jr commented:
The backwards compatibility section is in fact the only part required to be assigned a BIP number.

It should probably discuss compatibility with existing wallets, BIP 21, and BIP 70. Simply saying why there are no issues should be fine.

RHavar commented:
Great! Thanks for all the feedback so far. Updated.

luke-jr commented:
Assigned BIP 79, but note that the current title is too long (max length 44 chars)

luke-jr commented:
```
Invalid date format in bip-0079.mediawiki at scripts/buildtable.pl line 164, <$F> line 10.
```
